x/staking/keeper/cons_pubkey.go

```diff
@@ -0,0 +1,88 @@
+package keeper
+
+import (
+	"bytes"
+	"context"
+	"time"
+
+	"cosmossdk.io/collections"
+	"cosmossdk.io/x/staking/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// maxRotations is the value of max rotations can be made in unbonding period for a validator.
+const maxRotations = 1
+
+// setConsPubKeyRotationHistory sets the consensus key rotation of a validator into state
+func (k Keeper) setConsPubKeyRotationHistory(
+	ctx context.Context, valAddr sdk.ValAddress,
+	oldPubKey, newPubKey *codectypes.Any, fee sdk.Coin,
+) error {
+
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	height := uint64(sdkCtx.BlockHeight())
+	history := types.ConsPubKeyRotationHistory{
+		OperatorAddress: valAddr.Bytes(),
+		OldConsPubkey:   oldPubKey,
+		NewConsPubkey:   newPubKey,
+		Height:          height,
+		Fee:             fee,
+	}
+	err := k.RotationHistory.Set(ctx, valAddr, history)
+	if err != nil {
+		return err
+	}
+
+	ubdTime, err := k.UnbondingTime(ctx)
+	if err != nil {
+		return err
+	}
+
+	queueTime := sdkCtx.BlockHeader().Time.Add(ubdTime)
+	if err := k.ValidatorConsensusKeyRotationRecordIndexKey.Set(ctx, collections.Join(valAddr.Bytes(), queueTime), []byte{}); err != nil {
```

Comment:
if you add the `queueTime` as a field in `types.ConsPubKeyRotationHistory` this can also become a `MultiIndex`

Reply:
Hmm, but the record has to be in the state forever `queueTime` won't be helping much to store.

Reply:
this is fair, but I think making this an index brings some advantages, specifically less code and less required cognitive understanding of the separate state objects.

the extra queueTime field would be occupying only 12bytes of data (protobuf timestamp is 8bytes for seconds and 4bytes for nanos) – concretely speaking 83.000 rotations are required to cause a 1mb of bloat (extremely negligible). 

second, I think it should be reasonable to eventually cleanup rotation history, but this can be addressed later / separately if we assess this is safe.

Reply:
@atheeshp we can look at this later when we get the PR that makes use of the ValidatorConsensusKeyRotationRecordIndexKey.

For now I'd only suggest to make this object a KeySet since it appears to have no value.